src/settings/settinggroup.h

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef GROUPEDSETTINGS_H
+#define GROUPEDSETTINGS_H
+
+#include <QObject>
+#include <QVariant>
+
+#include "setting.h"
+
+/**
+ * @brief Represents a group of dynamic settings. Dynamic setttings can be
+ * defined at runtime.
+ *
+ * This abstraction serves as an aggregator of underlying storage settings.
+ * These settings necessarily need to be grouped under a QSettings group i.e.
+ * their keys need to share the same prefix followed by a slash ("/").
+ *
+ * Because settings under a SettingsGroup can be dynamically generated, upon
+ * initialization of a new SettingsGroup object the underlying storage is probed
+ * for any stored setting with the group prefix and those settings are
+ * registered to the SettingsBase.
+ *
+ * A SettingsGroups is useful when
+ *
+ * * Defining metrics dynamically at runtime is required or
+ * * Reacting to changes for _any_ setting under the group is necessary or
+ * * It makes sense to have metrics grouped for code semantics.
+ *
+ * It is allowed to create SettingGroups that are subsets or supersets of other
+ * SettingGroups. However, this can be tricky if the different groups are
+ * configured differently e.g. one has removeWhenReset set to true, while the
+ * other has removeWhenReset set to false. Whichever group register the
+ * underlying setting first will have their configurations enforced. There are
+ * Q_ASSERTs set up in debug mode to catch such situations.
+ *
+ */
+class SettingGroup : public QObject {
```

Comment:
Consider if moving the registration code into SettingsBase can provide a better design. (Better Single Responsibility). That may remove or reduce the requirement for the friend classes:

```
	  SettingGroup messageSettingGroup =
      SettingGroup(QString("%1/%2")
                       .arg(Constants::ADDONS_SETTINGS_GROUP)
                       .arg(ADDON_MESSAGE_SETTINGS_GROUP),
                   true,  // remove when reset
                   false  // sensitive setting
      );
      SettingsManager::instance()->register(messageSettingGroup);
```

Reply:
Would you mind clarifying where the registration you are referring to is right now? The SettingsGroup itself doesn't register anything, it's the SettingsFactory that registers settings in the SettingsBase.

Reply:
Sorry, I didn't look closely enough previously.

The friend classes in SettingsManager.h does break its encapsulation and makes it more difficult to understand the relationship between the classes. Consider removing those and introducing public methods which can be used by Setting and SetringGroup, instead of directly manipulating SettingManager's m_settings.

For example, introduce SettingsManager.getKeysInGroup(const QString& groupName), which can be used to get all those keys and add a slot to detect any changes in the group.

Reply:
Yeah ok... I learned about friend classes while working in this PR and maybe I got too excited. I'll do that.